# Predicting EUR/USD exchange rate

This notebook model predicts EUR/USD exchange rate using the pervious EUR pricing and sentiment around EUR and USD obtained from Twitter.

## Loading the data

Set the working directory and load eur pricing along with Twitter sentiment about Euro and US dollar.

In [1]:
import os

os.chdir("/media/dftortosa/Windows/Users/dftor/Documents/diego_docs/industry/data_incubator/capstone_project/")
os.getcwd()

'/media/dftortosa/Windows/Users/dftor/Documents/diego_docs/industry/data_incubator/capstone_project'

In [2]:
import pandas as pd
import numpy as np

eur_pricing = pd.read_csv("results/eur_pricing_data.csv.gz", compression="gzip")
eur_pricing

date  eur_usd_rate  previous_eur_pricing  \
0     1999-06-21        1.0339                1.0388   
1     1999-06-22        1.0313                1.0339   
2     1999-06-23        1.0280                1.0313   
3     1999-06-24        1.0321                1.0280   
4     1999-06-25        1.0443                1.0321   
...          ...           ...                   ...   
5933  2022-08-22        1.0001                1.0054   
5934  2022-08-23        0.9927                1.0001   
5935  2022-08-24        0.9934                0.9927   
5936  2022-08-25        0.9970                0.9934   
5937  2022-08-26        1.0007                0.9970   

      5_days_rolling_mean_pricing  10_days_rolling_mean_pricing  \
0                         1.03804                       1.04017   
1                         1.03608                       1.04040   
2                         1.03450                       1.03968   
3                         1.03320                       1.03782   
4                         1.03282                       1.03629   
...                           ...                           ...   
5933                      1.01444                       1.02030   
5934                      1.01056                       1.01832   
5935                      1.00648                       1.01525   
5936                      1.00188                       1.01207   
5937                      0.99772                       1.00839   

      60_days_rolling_mean_pricing  120_days_rolling_mean_pricing  \
0                         1.060603                       1.092362   
1                         1.060015                       1.091153   
2                         1.059352                       1.089922   
3                         1.058582                       1.088703   
4                         1.057830                       1.087611   
...                            ...                            ...   
5933                      1.034628                       1.057292   
5934                      1.033357                       1.056397   
5935                      1.032047                       1.055562   
5936                      1.030750                       1.054761   
5937                      1.029547                       1.053992   

      5_days_exp_rolling_mean_pricing  10_days_exp_rolling_mean_pricing  \
0                            1.038175                          1.040002   
1                            1.036750                          1.038892   
2                            1.034933                          1.037512   
3                            1.032622                          1.035782   
4                            1.032448                          1.035113   
...                               ...                               ...   
5933                         1.014329                          1.017477   
5934                         1.009586                          1.014317   
5935                         1.003957                          1.010387   
5936                         1.000438                          1.007298   
5937                         0.999292                          1.005426   

      60_days_exp_rolling_mean_pricing  120_days_exp_rolling_mean_pricing  
0                             1.062343                           1.087234  
1                             1.061410                           1.086353  
2                             1.060423                           1.085443  
3                             1.059360                           1.084493  
4                             1.058466                           1.083627  
...                                ...                                ...  
5933                          1.033201                           1.055635  
5934                          1.032116                           1.054717  
5935                          1.030824                           1.053692  
5936                          1.029597      

In [3]:
twitter_sentiment_eur = pd.read_csv("results/twitter_sentiment_data_eur.csv.gz", compression="gzip")
twitter_sentiment_eur

date  quantile_0.05_sent_eur_previous_1_days  \
0     2008-10-09                               -0.464215   
1     2008-10-10                               -0.421500   
2     2008-10-11                               -0.566100   
3     2008-10-12                               -0.443665   
4     2008-11-24                               -0.263670   
...          ...                                     ...   
4967  2022-08-20                               -0.817600   
4968  2022-08-21                               -0.840425   
4969  2022-08-22                               -0.848100   
4970  2022-08-23                               -0.825720   
4971  2022-08-24                               -0.822545   

      quantile_0.1_sent_eur_previous_1_days  \
0                                  -0.27320   
1                                  -0.42150   
2                                  -0.56030   
3                                  -0.12233   
4                                  -0.25414   
...                                     ...   
4967                               -0.73698   
4968                               -0.76145   
4969                               -0.79559   
4970                               -0.70858   
4971                               -0.72929   

      quantile_0.2_sent_eur_previous_1_days  \
0                                  -0.14796   
1                                  -0.37326   
2                                  -0.53898   
3                                   0.25262   
4                                  -0.23508   
...                                     ...   
4967                               -0.54230   
4968                               -0.59600   
4969                               -0.62490   
4970                               -0.55438   
4971                               -0.57190   

      quantile_0.3_sent_eur_previous_1_days  \
0                                   0.00000   
1                                  -0.28336   
2                                  -0.52670   
3                                   0.40924   
4                                  -0.21602   
...                                     ...   
4967                               -0.32233   
4968                               -0.37150   
4969                               -0.44865   
4970                               -0.36120   
4971                               -0.37921   

      quantile_0.4_sent_eur_previous_1_days  \
0                                   0.16630   
1                                  -0.10000   
2                                  -0.52670   
3                                   0.46062   
4                                  -0.19696   
...                                     ...   
4967                               -0.07720   
4968                               -0.15480   
4969                               -0.23486   
4970                               -0.12800   
4971                               -0.07720   

      quantile_0.5_sent_eur_previous_1_days  \
0                                    0.2263   
1                                    0.0000   
2                                   -0.5267   
3                                    0.5010   
4                                   -0.1779   
...                                     ...   
4967                                 0.0000   
4968                                 0.0000   
4969                                 0.0000   
4970                                 0.0000   
4971                                 0.0000   

      quantile_0.6_sent_eur_previous_1_days  \
0                                   0.22630   
1                                   0.00000   
2                                  -0.52670   
3                                   0.52670   
4                                  -0.14232   
...                                     ...   
4967                                0.17790   
4968                                0.12800   
4969                                0.03332

In [4]:
twitter_sentiment_usd = pd.read_csv("results/twitter_sentiment_data_usd.csv.gz", compression="gzip")
twitter_sentiment_usd

date  quantile_0.05_sent_usd_previous_1_days  \
0     2009-06-08                               -0.391100   
1     2009-06-09                               -0.526700   
2     2009-06-10                               -0.326940   
3     2009-06-11                               -0.725885   
4     2009-06-12                               -0.376910   
...          ...                                     ...   
4792  2022-08-20                               -0.736730   
4793  2022-08-21                               -0.802420   
4794  2022-08-22                               -0.659880   
4795  2022-08-23                               -0.755980   
4796  2022-08-24                               -0.752100   

      quantile_0.1_sent_usd_previous_1_days  \
0                                  -0.32770   
1                                  -0.52670   
2                                  -0.10898   
3                                  -0.59397   
4                                  -0.24554   
...                                     ...   
4792                               -0.57896   
4793                               -0.62970   
4794                               -0.55641   
4795                               -0.63690   
4796                               -0.63860   

      quantile_0.2_sent_usd_previous_1_days  \
0                                  -0.25418   
1                                  -0.44330   
2                                   0.00000   
3                                  -0.33512   
4                                  -0.14400   
...                                     ...   
4792                               -0.38180   
4793                               -0.41006   
4794                               -0.42432   
4795                               -0.45880   
4796                               -0.40190   

      quantile_0.3_sent_usd_previous_1_days  \
0                                  -0.22518   
1                                  -0.28638   
2                                   0.00000   
3                                   0.00000   
4                                   0.00000   
...                                     ...   
4792                               -0.16130   
4793                               -0.20744   
4794                               -0.31302   
4795                               -0.25000   
4796                               -0.20230   

      quantile_0.4_sent_usd_previous_1_days  \
0                                  -0.17880   
1                                  -0.06364   
2                                   0.00000   
3                                   0.00000   
4                                   0.00000   
...                                     ...   
4792                                0.00000   
4793                                0.00000   
4794                               -0.12800   
4795                                0.00000   
4796                                0.00000   

      quantile_0.5_sent_usd_previous_1_days  \
0                                     -0.00   
1                                      0.00   
2                                     -0.00   
3                                      0.35   
4                                      0.00   
...                                     ...   
4792                                  -0.00   
4793                                  -0.00   
4794                                   0.00   
4795                                  -0.00   
4796                                  -0.00   

      quantile_0.6_sent_usd_previous_1_days  \
0                                   0.00000   
1                                   0.00000   
2                                   0.00000   
3                                   0.41328   
4                                   0.00000   
...                                     ...   
4792                                0.00000   
4793                                0.00000   
4794                                0.00000

Merge into one single dataset:

In [5]:
from functools import reduce

#merge them using reduce
df_merged = reduce(lambda left, right: pd.merge(left, right, on=['date'], how='inner'), [eur_pricing, twitter_sentiment_eur, twitter_sentiment_usd])
    #leave only those rows having date in the datasets of currency value and EUR/USD sentiment
    #https://stackoverflow.com/questions/44327999/python-pandas-merge-multiple-dataframes
df_merged
    #note that here the previous eur pricing can be different than the eur pricing of the previous row
    #because when merging we lose days without tweet sentiment, so there is no longer continuous days

date  eur_usd_rate  previous_eur_pricing  \
0     2009-06-08        1.3866                1.4177   
1     2009-06-09        1.3959                1.3866   
2     2009-06-10        1.4102                1.3959   
3     2009-06-11        1.3969                1.4102   
4     2009-06-12        1.4004                1.3969   
...          ...           ...                   ...   
3362  2022-08-18        1.0178                1.0164   
3363  2022-08-19        1.0054                1.0178   
3364  2022-08-22        1.0001                1.0054   
3365  2022-08-23        0.9927                1.0001   
3366  2022-08-24        0.9934                0.9927   

      5_days_rolling_mean_pricing  10_days_rolling_mean_pricing  \
0                         1.41874                       1.40710   
1                         1.41166                       1.40566   
2                         1.40608                       1.40617   
3                         1.40398                       1.40818   
4                         1.40146                       1.40931   
...                           ...                           ...   
3362                      1.02226                       1.02212   
3363                      1.01906                       1.02209   
3364                      1.01444                       1.02030   
3365                      1.01056                       1.01832   
3366                      1.00648                       1.01525   

      60_days_rolling_mean_pricing  120_days_rolling_mean_pricing  \
0                         1.344202                       1.329956   
1                         1.346002                       1.330394   
2                         1.347963                       1.330768   
3                         1.349958                       1.331112   
4                         1.351503                       1.331037   
...                            ...                            ...   
3362                      1.036607                       1.058989   
3363                      1.035742                       1.058169   
3364                      1.034628                       1.057292   
3365                      1.033357                       1.056397   
3366                      1.032047                       1.055562   

      5_days_exp_rolling_mean_pricing  10_days_exp_rolling_mean_pricing  \
0                            1.414169                          1.405729   
1                            1.404979                          1.402251   
2                            1.401953                          1.401096   
3                            1.404702                          1.402752   
4                            1.402101                          1.401688   
...                               ...                               ...   
3362                         1.019290                          1.020685   
3363                         1.018794                          1.020160   
3364                         1.014329                          1.017477   
3365                         1.009586                          1.014317   
3366                         1.003957                          1.010387   

      60_days_exp_rolling_mean_pricing  ...  \
0                             1.352455  ...   
1                             1.353575  ...   
2                             1.354963  ...   
3                             1.356774  ...   
4                             1.358089  ...   
...                                ...  ...   
3362                          1.034698  ...   
3363                          1.034144  ...   
3364                          1.033201  ...   
3365                          1.032116  ...   
3366                          1.030824  ...   

      quantile_0.95_sent_usd_previous_15_days  mean_sent_usd_previous_15_days  \
0                                    0.245880                        0.091067   
1                                    0.480525                        0.185367

Subset the data selecting only predictors, including EUR pricing, EUR and USD sentiment:

In [6]:
predictors = df_merged[[column for column in df_merged.columns if "sent_eur_previous" in column or "sent_usd_previous" in column or "previous_eur_pricing" in column or "rolling_mean" in column]]
predictors

previous_eur_pricing  5_days_rolling_mean_pricing  \
0                   1.4177                      1.41874   
1                   1.3866                      1.41166   
2                   1.3959                      1.40608   
3                   1.4102                      1.40398   
4                   1.3969                      1.40146   
...                    ...                          ...   
3362                1.0164                      1.02226   
3363                1.0178                      1.01906   
3364                1.0054                      1.01444   
3365                1.0001                      1.01056   
3366                0.9927                      1.00648   

      10_days_rolling_mean_pricing  60_days_rolling_mean_pricing  \
0                          1.40710                      1.344202   
1                          1.40566                      1.346002   
2                          1.40617                      1.347963   
3                          1.40818                      1.349958   
4                          1.40931                      1.351503   
...                            ...                           ...   
3362                       1.02212                      1.036607   
3363                       1.02209                      1.035742   
3364                       1.02030                      1.034628   
3365                       1.01832                      1.033357   
3366                       1.01525                      1.032047   

      120_days_rolling_mean_pricing  5_days_exp_rolling_mean_pricing  \
0                          1.329956                         1.414169   
1                          1.330394                         1.404979   
2                          1.330768                         1.401953   
3                          1.331112                         1.404702   
4                          1.331037                         1.402101   
...                             ...                              ...   
3362                       1.058989                         1.019290   
3363                       1.058169                         1.018794   
3364                       1.057292                         1.014329   
3365                       1.056397                         1.009586   
3366                       1.055562                         1.003957   

      10_days_exp_rolling_mean_pricing  60_days_exp_rolling_mean_pricing  \
0                             1.405729                          1.352455   
1                             1.402251                          1.353575   
2                             1.401096                          1.354963   
3                             1.402752                          1.356774   
4                             1.401688                          1.358089   
...                                ...                               ...   
3362                          1.020685                          1.034698   
3363                          1.020160                          1.034144   
3364                          1.017477                          1.033201   
3365                          1.014317                          1.032116   
3366                          1.010387                          1.030824   

      120_days_exp_rolling_mean_pricing  \
0                              1.344919   
1                              1.345608   
2                              1.346439   
3                              1.347493   
4                              1.348310   
...                                 ...   
3362                           1.057130   
3363                           1.056479   
3364                           1.055635   
3365                           1.054717   
3366                           1.053692   

      quantile_0.05_sent_eur_previous_1_days  ...  \
0                                  -0.593980  ...   
1                                  -0.868900  ...   
2                         

Select only those predictors related to EUR pricing:

In [7]:
predictors_eur_pricing = predictors[[column for column in predictors.columns if "rolling_mean_" in column or "previous_eur_pricing" in column]]
predictors_eur_pricing

previous_eur_pricing  5_days_rolling_mean_pricing  \
0                   1.4177                      1.41874   
1                   1.3866                      1.41166   
2                   1.3959                      1.40608   
3                   1.4102                      1.40398   
4                   1.3969                      1.40146   
...                    ...                          ...   
3362                1.0164                      1.02226   
3363                1.0178                      1.01906   
3364                1.0054                      1.01444   
3365                1.0001                      1.01056   
3366                0.9927                      1.00648   

      10_days_rolling_mean_pricing  60_days_rolling_mean_pricing  \
0                          1.40710                      1.344202   
1                          1.40566                      1.346002   
2                          1.40617                      1.347963   
3                          1.40818                      1.349958   
4                          1.40931                      1.351503   
...                            ...                           ...   
3362                       1.02212                      1.036607   
3363                       1.02209                      1.035742   
3364                       1.02030                      1.034628   
3365                       1.01832                      1.033357   
3366                       1.01525                      1.032047   

      120_days_rolling_mean_pricing  5_days_exp_rolling_mean_pricing  \
0                          1.329956                         1.414169   
1                          1.330394                         1.404979   
2                          1.330768                         1.401953   
3                          1.331112                         1.404702   
4                          1.331037                         1.402101   
...                             ...                              ...   
3362                       1.058989                         1.019290   
3363                       1.058169                         1.018794   
3364                       1.057292                         1.014329   
3365                       1.056397                         1.009586   
3366                       1.055562                         1.003957   

      10_days_exp_rolling_mean_pricing  60_days_exp_rolling_mean_pricing  \
0                             1.405729                          1.352455   
1                             1.402251                          1.353575   
2                             1.401096                          1.354963   
3                             1.402752                          1.356774   
4                             1.401688                          1.358089   
...                                ...                               ...   
3362                          1.020685                          1.034698   
3363                          1.020160                          1.034144   
3364                          1.017477                          1.033201   
3365                          1.014317                          1.032116   
3366                          1.010387                          1.030824   

      120_days_exp_rolling_mean_pricing  
0                              1.344919  
1                              1.345608  
2                              1.346439  
3                              1.347493  
4                              1.348310  
...                                 ...  
3362                           1.057130  
3363                           1.056479  
3364                           1.055635  
3365                           1.054717  
3366                           1.053692  

[3367 rows x 9 columns]

Select predictors related to Twitter sentiment:

In [8]:
predictors_twitter = predictors[[column for column in predictors.columns if "previous_eur_pricing" not in column and "rolling_mean" not in column]]
predictors_twitter

quantile_0.05_sent_eur_previous_1_days  \
0                                  -0.593980   
1                                  -0.868900   
2                                  -0.664245   
3                                  -0.860135   
4                                  -0.930000   
...                                      ...   
3362                               -0.831600   
3363                               -0.798310   
3364                               -0.848100   
3365                               -0.825720   
3366                               -0.822545   

      quantile_0.1_sent_eur_previous_1_days  \
0                                  -0.45286   
1                                  -0.86890   
2                                  -0.42874   
3                                  -0.73678   
4                                  -0.92862   
...                                     ...   
3362                               -0.70970   
3363                               -0.67050   
3364                               -0.79559   
3365                               -0.70858   
3366                               -0.72929   

      quantile_0.2_sent_eur_previous_1_days  \
0                                  -0.42150   
1                                  -0.48648   
2                                   0.00000   
3                                  -0.47670   
4                                  -0.83328   
...                                     ...   
3362                               -0.54230   
3363                               -0.45636   
3364                               -0.62490   
3365                               -0.55438   
3366                               -0.57190   

      quantile_0.3_sent_eur_previous_1_days  \
0                                  -0.37929   
1                                   0.00000   
2                                   0.00000   
3                                  -0.40959   
4                                  -0.50286   
...                                     ...   
3362                               -0.27390   
3363                               -0.27320   
3364                               -0.44865   
3365                               -0.36120   
3366                               -0.37921   

      quantile_0.4_sent_eur_previous_1_days  \
0                                  -0.14448   
1                                   0.00000   
2                                   0.00000   
3                                  -0.28688   
4                                  -0.05160   
...                                     ...   
3362                                0.00000   
3363                                0.00000   
3364                               -0.23486   
3365                               -0.12800   
3366                               -0.07720   

      quantile_0.5_sent_eur_previous_1_days  \
0                                       0.0   
1                                       0.0   
2                                       0.0   
3                                       0.0   
4                                       0.0   
...                                     ...   
3362                                    0.0   
3363                                    0.0   
3364                                    0.0   
3365                                    0.0   
3366                                    0.0   

      quantile_0.6_sent_eur_previous_1_days  \
0                                   0.01032   
1                                   0.00000   
2                                   0.00000   
3                                   0.00000   
4                                   0.10270   
...                                     ...   
3362                                0.17790   
3363                                0.17790   
3364                                0.03332   
3365                                0.03186   
3366                                0.14698   

      quantile_0.7_sent_eur_previous_1_

Convert the response variable, which is the EUR/USD exchange rate, and the predictors to numpy arrays:

In [9]:
y = df_merged["eur_usd_rate"].values
X = predictors.values #all predictors
X_eur_pricing = predictors_eur_pricing.values #eur pricing predictors
X_twitter = predictors_twitter.values #twitter sentiment predictors

In [10]:
print(y.shape)
print(X.shape)
print(X_eur_pricing.shape)
print(X_twitter.shape)

(3367,)
(3367, 609)
(3367, 9)
(3367, 600)


In [11]:
f"columns in X are a combination of X_eur_pricing and X_twitter? {X.shape[1] == X_eur_pricing.shape[1] + X_twitter.shape[1]}"

'columns in X are a combination of X_eur_pricing and X_twitter? True'

## Compare regression models with default parameters

Comparison of mulitple models recommended for regression problems with less than 100K observations ([scikit learn flowchart](https://scikit-learn.org/stable/tutorial/machine_learning_map/index.html)). In each case, three models will be fitted, a full model with all predictors, a model with only EUR pricing predictors and a model with only Twitter sentiment predictors.

First, calculate the predictive power (R<sup>2</sup>) models using Cross Validation. In particular, using ShuffleSplit in order to create random splits for training and test sets. These splits will be used in `cross_val_score` to calculate the R<sup>2</sup> in each test set.

In [12]:
from sklearn.model_selection import ShuffleSplit
shuffle_split = ShuffleSplit(n_splits=5, test_size=0.3, random_state=1)

### Regression models for few important features

Given the lack of previous knowledge about whether one or multiple Twitter predictors should be relevant in the prediction, models with better performance in both scenarios will be tested. This section will show those recommended for cases with just a few important features:

#### Lasso

In [13]:
from sklearn.linear_model import Lasso
from sklearn.model_selection import cross_val_score
print("R2 for the full model:")
np.mean(cross_val_score(estimator=Lasso(), 
                        X=X, 
                        y=y, 
                        cv=shuffle_split, 
                        scoring="r2",
                        n_jobs=5))

R2 for the full model:


0.39566469945020427

In [14]:
print("R2 for the EUR-only model:")
np.mean(cross_val_score(estimator=Lasso(), 
                        X=X_eur_pricing, 
                        y=y, 
                        cv=shuffle_split, 
                        scoring="r2",
                        n_jobs=5))

R2 for the EUR-only model:


-0.0007626642910206094

In [15]:
print("R2 for the Twitter-only model:")
np.mean(cross_val_score(estimator=Lasso(), 
                        X=X_twitter, 
                        y=y, 
                        cv=shuffle_split, 
                        scoring="r2",
                        n_jobs=5))

R2 for the Twitter-only model:


0.39566469945020427

#### Elastic net

In [16]:
from sklearn.linear_model import ElasticNet
print("R2 for the full model:")
np.mean(cross_val_score(estimator=ElasticNet(), 
                        X=X, 
                        y=y, 
                        cv=shuffle_split, 
                        scoring="r2",
                        n_jobs=5))

R2 for the full model:


0.4339773097125318

In [17]:
print("R2 for the EUR-only model:")
np.mean(cross_val_score(estimator=ElasticNet(), 
                        X=X_eur_pricing, 
                        y=y, 
                        cv=shuffle_split, 
                        scoring="r2",
                        n_jobs=5))

R2 for the EUR-only model:


-0.0007626642910206094

In [18]:
print("R2 for the Twitter-only model:")
np.mean(cross_val_score(estimator=ElasticNet(), 
                        X=X_twitter, 
                        y=y, 
                        cv=shuffle_split, 
                        scoring="r2",
                        n_jobs=5))

R2 for the Twitter-only model:


0.4339773097125318

### Multiple important features

#### Ridge

In [19]:
from sklearn.linear_model import Ridge
print("R2 for the full model:")
np.mean(cross_val_score(estimator=Ridge(), 
                        X=X, 
                        y=y, 
                        cv=shuffle_split, 
                        scoring="r2",
                        n_jobs=5))

R2 for the full model:


0.9927989059836312

In [20]:
print("R2 for the EUR-only model:")
np.mean(cross_val_score(estimator=Ridge(), 
                        X=X_eur_pricing, 
                        y=y, 
                        cv=shuffle_split, 
                        scoring="r2",
                        n_jobs=5))

R2 for the EUR-only model:


0.994532862029738

In [21]:
print("R2 for the Twitter-only model:")
np.mean(cross_val_score(estimator=Ridge(), 
                        X=X_twitter, 
                        y=y, 
                        cv=shuffle_split, 
                        scoring="r2",
                        n_jobs=5))

R2 for the Twitter-only model:


0.658829653121915

#### Supporting Vector Machines linear

In [ ]:
from sklearn.svm import SVR
print("R2 for the full model:")
np.mean(cross_val_score(estimator=SVR(kernel="linear"), 
                        X=X, 
                        y=y, 
                        cv=shuffle_split, 
                        scoring="r2",
                        n_jobs=5))

R2 for the full model:


In [ ]:
print("R2 for the EUR-only model:")
np.mean(cross_val_score(estimator=SVR(kernel="linear"), 
                        X=X_eur_pricing, 
                        y=y, 
                        cv=shuffle_split, 
                        scoring="r2",
                        n_jobs=5))

In [ ]:
print("R2 for the Twitter-only model:")
np.mean(cross_val_score(estimator=SVR(kernel="linear"), 
                        X=X_twitter, 
                        y=y, 
                        cv=shuffle_split, 
                        scoring="r2",
                        n_jobs=5))

#### Supporting Vector Machines non-linear

In [22]:
from sklearn.svm import SVR
print("R2 for the full model:")
np.mean(cross_val_score(estimator=SVR(kernel="rbf"), 
                        X=X, 
                        y=y, 
                        cv=shuffle_split, 
                        scoring="r2",
                        n_jobs=5))

R2 for the full model:


0.5682845612266301

In [23]:
print("R2 for the EUR-only model:")
np.mean(cross_val_score(estimator=SVR(kernel="rbf"), 
                        X=X_eur_pricing, 
                        y=y, 
                        cv=shuffle_split, 
                        scoring="r2",
                        n_jobs=5))

R2 for the EUR-only model:


0.8859508822048323

In [24]:
print("R2 for the Twitter-only model:")
np.mean(cross_val_score(estimator=SVR(kernel="rbf"), 
                        X=X_twitter, 
                        y=y, 
                        cv=shuffle_split, 
                        scoring="r2",
                        n_jobs=5))

R2 for the Twitter-only model:


0.5682845791597761

#### Ensemble regressors

##### Random Forest

In [25]:
from sklearn.ensemble import RandomForestRegressor
print("R2 for the full model:")
np.mean(cross_val_score(estimator=RandomForestRegressor(random_state=1), 
                        X=X, 
                        y=y, 
                        cv=shuffle_split, 
                        scoring="r2",
                        n_jobs=5))

R2 for the full model:


0.9965666025378589

In [26]:
print("R2 for the EUR-only model:")
np.mean(cross_val_score(estimator=RandomForestRegressor(random_state=1),
                        X=X_eur_pricing, 
                        y=y, 
                        cv=shuffle_split, 
                        scoring="r2",
                        n_jobs=5))

R2 for the EUR-only model:


0.9964007207530393

In [27]:
print("R2 for the Twitter-only model:")
np.mean(cross_val_score(estimator=RandomForestRegressor(random_state=1),
                        X=X_twitter, 
                        y=y, 
                        cv=shuffle_split, 
                        scoring="r2",
                        n_jobs=5))

R2 for the Twitter-only model:


0.8153337261409161

##### Extra Trees Regressor

In [28]:
from sklearn.ensemble import ExtraTreesRegressor
print("R2 for the full model:")
np.mean(cross_val_score(estimator=ExtraTreesRegressor(random_state=1), 
                        X=X, 
                        y=y,
                        cv=shuffle_split, 
                        scoring="r2",
                        n_jobs=5))

R2 for the full model:


0.9966581052766154

In [29]:
print("R2 for the EUR-only model:")
np.mean(cross_val_score(estimator=ExtraTreesRegressor(random_state=1),
                        X=X_eur_pricing, 
                        y=y, 
                        cv=shuffle_split, 
                        scoring="r2",
                        n_jobs=5))

R2 for the EUR-only model:


0.9964122024670665

In [30]:
print("R2 for the Twitter-only model:")
np.mean(cross_val_score(estimator=ExtraTreesRegressor(random_state=1),
                        X=X_twitter, 
                        y=y, 
                        cv=shuffle_split, 
                        scoring="r2",
                        n_jobs=5))

R2 for the Twitter-only model:


0.8302671551647759

##### Gradient Boost

In [31]:
from sklearn.ensemble import GradientBoostingRegressor
print("R2 for the full model:")
np.mean(cross_val_score(estimator=GradientBoostingRegressor(random_state=1), 
                        X=X, 
                        y=y, 
                        cv=shuffle_split, 
                        scoring="r2",
                        n_jobs=5))

R2 for the full model:


0.9965579999438605

In [32]:
print("R2 for the EUR-only model:")
np.mean(cross_val_score(estimator=GradientBoostingRegressor(random_state=1),
                        X=X_eur_pricing, 
                        y=y, 
                        cv=shuffle_split, 
                        scoring="r2",
                        n_jobs=5))

R2 for the EUR-only model:


0.996630021114149

In [33]:
print("R2 for the Twitter-only model:")
np.mean(cross_val_score(estimator=GradientBoostingRegressor(random_state=1),
                        X=X_twitter, 
                        y=y, 
                        cv=shuffle_split, 
                        scoring="r2",
                        n_jobs=5))

R2 for the Twitter-only model:


0.802182601552715

##### Voting regressor

Define first the models that will be included in the votting regressor:

In [34]:
from sklearn.ensemble import VotingRegressor
reg1 = GradientBoostingRegressor(random_state=1)
reg2 = ExtraTreesRegressor(random_state=1)
reg3 = Ridge()

In [35]:
print("R2 for the full model:")
np.mean(cross_val_score(estimator=VotingRegressor(estimators=[('gb', reg1), ('et', reg2), ('rdg', reg3)]), 
                        X=X, 
                        y=y, 
                        cv=shuffle_split, 
                        scoring="r2",
                        n_jobs=5))

R2 for the full model:


0.9961732282481621

In [36]:
print("R2 for the EUR-only model:")
np.mean(cross_val_score(estimator=VotingRegressor(estimators=[('gb', reg1), ('et', reg2), ('rdg', reg3)]),
                        X=X_eur_pricing, 
                        y=y, 
                        cv=shuffle_split, 
                        scoring="r2",
                        n_jobs=5))

R2 for the EUR-only model:


0.9964867610793011

In [37]:
print("R2 for the Twitter-only model:")
np.mean(cross_val_score(estimator=VotingRegressor(estimators=[('gb', reg1), ('et', reg2), ('rdg', reg3)]),
                        X=X_twitter, 
                        y=y, 
                        cv=shuffle_split, 
                        scoring="r2",
                        n_jobs=5))

R2 for the Twitter-only model:


0.8044995635266016

The higest R<sup>2</sup> was found for the full model in Extra Trees, so we will use this model in next steps.

## Extra Trees Regressor

### Grid search

Perform a random grid search for the three different models in order to optimize the hyperparameters in each one.

In [38]:
full_model_grid_search = ExtraTreesRegressor(random_state=1)
eur_model_grid_search = ExtraTreesRegressor(random_state=1)
twitter_model_grid_search = ExtraTreesRegressor(random_state=1)

Prepare a dict with the parameters:

In [39]:
##parameters
# Number of trees in random forest
n_estimators = [100, 1000, 2000, 3000, 4000]
# Number of features to consider at every split
max_features = ['auto', 'sqrt', 'log']
# Maximum number of levels in tree
max_depth = [10, 100, 200, 300, 400, 500]
max_depth.append(None)
# Minimum number of samples required to split a node
min_samples_split = [2, 5, 10, 50, 70, 100]
# Minimum number of samples required at each leaf node
min_samples_leaf = [1, 2, 10, 40, 80, 100, 200, 300, 400]
# Method of selecting samples for training each tree
bootstrap = [True, False]

#make a dict with the parameters and values to be explored
params = {'n_estimators': n_estimators,
               'max_features': max_features,
               'max_depth': max_depth,
               'min_samples_split': min_samples_split,
               'min_samples_leaf': min_samples_leaf,
               'bootstrap': bootstrap}

Prepare the random search for each of the three models:

In [40]:
from sklearn.model_selection import RandomizedSearchCV

search_full_model = RandomizedSearchCV(estimator=full_model_grid_search, 
                                       param_distributions=params,
                                       n_iter=100,
                                       cv=shuffle_split, 
                                       verbose=2,
                                       n_jobs=-2, 
                                       scoring="r2")
search_eur_model = RandomizedSearchCV(estimator=eur_model_grid_search, 
                                       param_distributions=params,
                                       n_iter=100,
                                       cv=shuffle_split, 
                                       verbose=2,
                                       n_jobs=-2, 
                                       scoring="r2")
search_twitter_model = RandomizedSearchCV(estimator=twitter_model_grid_search, 
                                       param_distributions=params,
                                       n_iter=100,
                                       cv=shuffle_split, 
                                       verbose=2,
                                       n_jobs=-2, 
                                       scoring="r2")

Run the search:

In [ ]:
#make the search for out data
search_full_model.fit(X, y)
search_eur_model.fit(X_eur_pricing, y)
search_twitter_model.fit(X_twitter, y)

Fitting 5 folds for each of 100 candidates, totalling 500 fits


/home/dftortosa/.local/lib/python3.6/site-packages/sklearn/model_selection/_search.py:925: UserWarning: One or more of the test scores are non-finite: [       nan 0.97021743 0.92696612 0.99491733 0.93092767 0.99491219
 0.84380417 0.72709733        nan 0.86948122 0.92879301 0.9235525
 0.99491377 0.99025893        nan        nan        nan        nan
 0.9752406  0.95769454 0.91056541        nan 0.78636665 0.99468361
 0.99326835        nan        nan        nan 0.99648382        nan
        nan 0.75873454 0.91077496 0.68755576 0.83122565        nan
        nan 0.68755576 0.97969581 0.48590191 0.97332329 0.98575866
 0.9431909  0.99654172 0.99620795 0.99326389 0.93359722 0.9326563
 0.99618094 0.97010608 0.99491733 0.58070973 0.72885213 0.76578605
 0.99031818 0.99179758 0.98592952 0.99654594 0.52634741 0.91737731
        nan 0.99621023        nan        nan        nan        nan
 0.99621033 0.95773732 0.99605013 0.9431909  0.9602117         nan
 0.93726784 0.99179866 0.46595139 0.95032427 0.

Fitting 5 folds for each of 100 candidates, totalling 500 fits


/home/dftortosa/.local/lib/python3.6/site-packages/sklearn/model_selection/_search.py:925: UserWarning: One or more of the test scores are non-finite: [0.99648472 0.95583072 0.9898092  0.96331772 0.98545118 0.84915612
        nan 0.99492662 0.90904288 0.99273512        nan        nan
 0.98113706 0.97167411 0.99667003        nan 0.96242255 0.99620984
        nan 0.8786028         nan 0.95306888 0.99619486 0.98958205
 0.99629649 0.99565524 0.98495089 0.99633607 0.99276185 0.9911372
 0.99667983        nan        nan        nan        nan 0.9911372
 0.95860982 0.84754561        nan        nan 0.98003638 0.9966799
 0.98549761 0.97149201        nan 0.88164115        nan 0.98533114
 0.99648305        nan 0.8786028  0.90648457 0.98113706 0.98820047
 0.99415739 0.99220542 0.9582896  0.91039914        nan 0.98958205
        nan        nan 0.93668925 0.97873704 0.93791371 0.97191367
 0.98980984 0.99655974 0.98517421        nan 0.9957048  0.99661034
 0.93803458 0.93791371 0.90599209        nan 0.8

Fitting 5 folds for each of 100 candidates, totalling 500 fits


See the best parameters for each model:

In [ ]:
#print the best score with 3 decimals
print("Best score full model (CV score=%0.3f):" % search_full_model.best_score_)
print("Best score eur pricing model (CV score=%0.3f):" % search_eur_model.best_score_)
print("Best score twitter model (CV score=%0.3f):" % search_twitter_model.best_score_)

In [ ]:
#print the best value of the hyperparameter
print("Best parameters full model")
print(search_full_model.best_params_)
print("Best parameters EUR-only model")
print(search_eur_model.best_params_)
print("Best parameters Twitter-only model")
print(search_twitter_model.best_params_)

In [ ]:
#search_full_model.cv_results_
#search_eur_model.cv_results_
#search_twitter_model.cv_results_

### Optimized models

Check the performance of the models with the optimized parameters

In [ ]:
full_model = RandomForestRegressor(n_estimators=2000, 
                                   min_samples_split=5, 
                                   min_samples_leaf=2, 
                                   max_features="auto",
                                   max_depth=50,
                                   bootstrap=True, 
                                   random_state=1)
eur_model = RandomForestRegressor(n_estimators=1400, 
                                   min_samples_split=5, 
                                   min_samples_leaf=1, 
                                   max_features="sqrt",
                                   max_depth=30,
                                   bootstrap=True, 
                                   random_state=1)
twitter_model = RandomForestRegressor(n_estimators=800, 
                                   min_samples_split=2, 
                                   min_samples_leaf=1, 
                                   max_features="sqrt",
                                   max_depth=20,
                                   bootstrap=False, 
                                   random_state=1)

In [ ]:
print("R2 for the full model:")
np.mean(cross_val_score(estimator=full_model, 
                        X=X, 
                        y=y, 
                        cv=shuffle_split, 
                        scoring="r2",
                        n_jobs=5))

In [ ]:
print("R2 for the EUR-only model:")
np.mean(cross_val_score(estimator=eur_model, 
                        X=X_eur_pricing, 
                        y=y, 
                        cv=shuffle_split, 
                        scoring="r2",
                        n_jobs=5))

In [ ]:
print("R2 for the Twitter-only model:")
np.mean(cross_val_score(estimator=twitter_model, 
                        X=X_twitter, 
                        y=y, 
                        cv=shuffle_split, 
                        scoring="r2",
                        n_jobs=5))

Now fit and predict in the whole dataset:

In [ ]:
full_model.fit(X=X, y=y)
eur_model.fit(X=X_eur_pricing, y=y)
twitter_model.fit(X=X_twitter, y=y)

In [ ]:
full_model_pred = full_model.predict(X=X)
eur_model_pred = eur_model.predict(X=X_eur_pricing)
twitter_model_pred = twitter_model.predict(X=X_twitter)

See predictors importance:

In [ ]:
#full model
pd.Series(full_model.feature_importances_, index=predictors.columns).nlargest(10).plot(kind='barh')

In [ ]:
#eur pricing model only
pd.Series(eur_model.feature_importances_, index=predictors_eur_pricing.columns).nlargest(10).plot(kind='barh')

In [ ]:
#twitter predictors only
pd.Series(twitter_model.feature_importances_, index=predictors_twitter.columns).nlargest(10).plot(kind='barh')

In [ ]:
import matplotlib.pyplot as plt
from sklearn import metrics 

#open a plot with two panels
fig, (ax1, ax2, ax3) = plt.subplots(1,3, figsize=(15,4))

#plot scatter against date and eur pricing
ax1.scatter(x=full_model_pred, y=y, s=0.5)
ax2.scatter(x=eur_model_pred, y=y, s=0.5)
ax3.scatter(x=twitter_model_pred, y=y, s=0.5)

#set titles and axes labels
ax1.title.set_text("Full model")
ax1.set_ylabel("Observed EUR/USD exchange")
ax1.set_xlabel("Predicted EUR/USD exchange")
ax2.title.set_text("EUR-only model")
ax2.set_xlabel("Predicted EUR/USD exchange")
ax3.title.set_text("Twitter-only model")
ax3.set_xlabel("Predicted EUR/USD exchange")


#add R2
ax1.text(0.40, 0.8, 'R2 (whole dataset) = ' + str(np.round(metrics.r2_score(y, full_model_pred), 6)), horizontalalignment='center', verticalalignment='center', transform=ax1.transAxes)
ax2.text(1.61, 0.8, 'R2 (whole dataset) = ' + str(np.round(metrics.r2_score(y, eur_model_pred), 6)), horizontalalignment='center', verticalalignment='center', transform=ax1.transAxes)
ax3.text(2.75, 0.8, 'R2 (whole dataset) = ' + str(np.round(metrics.r2_score(y, twitter_model_pred), 6)), horizontalalignment='center', verticalalignment='center', transform=ax1.transAxes)

#close
plt.show()

The Twitter-only model works best than the other two, but note that this includes the whole dataset and no CV was perfomed in order to get these R<sup>2</sup>.

## Simplest model

Model considering the EUR pricing (EUR/USD exchange rate) of the previous day as the value of the current day: 

In [ ]:
metrics.r2_score(y, df_merged["previous_eur_pricing"].values)

In [ ]:
#open a plot with two panels
fig, ax1 = plt.subplots(1,1, figsize=(15,4))

#plot scatter against date and eur pricing
ax1.scatter(x=df_merged["previous_eur_pricing"], y=y, s=0.5)

The price of Euro seems to be very stable, it can change from 1.2 to 1.1 but with small steps. This explains why the simplest model with just the previous EUR pricing without any modeling approach gets the highest R<sup>2</sup>.

Extra Tree Regression models surpass this, being the Twitter-only model the one with the highest predcitive power. However, when using CV to calculate R<sup>2</sup>, predictive power is a little bit lower in the extra tree regressors, being the highest for the full model (Twitter + EUR pricing). Despite this, it is relevant the fact that models including Twitter information work better in general than the EUR-only model.

Therefore, it could be relevant the consideration of expectations around a fiat currency using Twitter, and it could be more relevant if the currency is less stable than the Euro. Note that here I considered Twitter sentiment of the previous 15 days, so it may be the case that rapid changes in the expectations around a less estable currency could be detected in twitter anticipating changes in the value of the currency.